### PR TITLE
A string composed of many digits can be a string

### PR DESCRIPTION
**Pull request by akeeman**
Let's discuss the string of digits versus big int thing.

The json schema test cases prescribe that `98249283749234923498293171823948729348710298301928331` should be interpreted as a number and implicitly that `"x8249283749234923498293171823948729348710298301928331"` is a string (note the leading '9' has been replaced by an 'x' and that it actually is a string). So far so good.

Then there is our big friend `JSON_BIGINT_AS_STRING`. Using it results in a case where the validator actually cannot tell if someone submitted the number `98249283749234923498293171823948729348710298301928331` or the string `"98249283749234923498293171823948729348710298301928331"`. That's where the hassle begins.

The [test suite](https://github.com/json-schema/JSON-Schema-Test-Suite) also prescribes in an optional test that a big int is not a string. Sounds logical. However, as I just described, PHP cannot tell the difference anymore between if the value was submitted as a string or that is has been submitted as an integer after decoding it with `JSON_BIGINT_AS_STRING`.
The choice made so far has been to check if a string is composed out of digits only and check if it represents a value bigger then `PHP_INT_MAX`. However, this results in the problem that no strings can be validated as string if they happen to be composed out of a lot of digits only.
I think that's wrong, because certain strings actually can be just strings and should be able to validate as such.

I also think that there are several ways to handle this case better of which I'm describing two below.
1. The first is the staring point of this pr. It simply lets lets the big number validate as a valid string too (in addition to validating it as a valid integer), and adds line in the tests to convert the 'big int is not a string' test into a 'big int is a string' test.
2. The second is a bit more bulky but maybe more correct: adding an option (constructor and/or setter) to the Type constraint to let the end user decide which mode should be used. Something like `$validator->getRuleSet()->get('type')->setBigIntMode(Type::BIG_INT_AS_STRING);`, where it would validate a big int and a string of digits as strings.

Let me know your thoughts! :)

### CODE REVIEW

**Comment by yuloh** [comment]
Thanks for the detailed PR 👍 

#1 is probably the most correct.  Allowing a number to pass which is actually a string once it's decoded is better than rejecting a valid string because it's numeric.  If we allow the big numeric strings the user can always add a `pattern` constraint to reject them.

#2 is nice because it's not a breaking change and can be released in a minor version, and we can change the default behavior in the next major version.

**Comment by akeeman** [comment]
I agree. So I think I can conclude that option 2 is the way to go, as it allows for smooth transition.
I'll revert the changes and implement that one.

**Comment by yuloh** [comment]
👍  Sounds good.  I've been adding both a constructor parameter and a  setter for constraint settings, since the setter is simpler to use if you are using the default rule set.

Thanks for your help!

**Comment by akeeman** [comment]
I was thinking something like that. But I'm not too happy with the named, though.

**Comment by akeeman** [comment]
Added a static parameter to be able to set a runtime-wide default value.

I think that finishes up this pr just nicely, don't you think?

**Comment by coolhome** [comment]
Just ran into this issue at work 😢 

```json
{
  "keyword": "type",
  "message": "Value \"101010101010\" is not a(n) \"string\"",
  "pointer": "/thing",
  "value": "101010101010",
  "context": {
    "value": "\"101010101010\"",
    "type": "\"string\""
  }
}
```

**Comment by akeeman** [comment]
@coolhome, please note that this pr is targeted at version 1.1 and you are probably running 0.5.x.

File: src/Constraint/DraftFour/Type.php [hunk 1]
```diff
@@ -13,6 +13,31 @@
...
+     * Whether examples like 98249283749234923498293171823948729348710298301928331
+     * and "98249283749234923498293171823948729348710298301928331" are valid strings.
+     */
+    const BIGINT_MODE_STRING_VALID = 1;
+    const BIGINT_MODE_STRING_INVALID = 2;
+
+    /**
+     * @var int
+     */
+    private static $defaultBigintMode = self::BIGINT_MODE_STRING_INVALID;
```
Comment (yuloh) [inline src/Constraint/DraftFour/Type.php]: The ruleset binds the instance as a singleton, so you should only have to set it once per ruleset instance.

```php
$ruleset = new DraftFour();

$ruleset->get(Type::KEYWORD)->setBigIntMode(TYPE::BIGINT_MODE_STRING_VALID);

$validator = new Validator($data, $schema, $ruleset);
$validator = new Validator($otherData, $otherSchema, $ruleset);
```

It's not as convenient as the static parameter but it's consistent with the the other constraints that accept configuration parameters.

Comment (akeeman) [inline src/Constraint/DraftFour/Type.php]: Sounds fair :+1:

File: src/Constraint/DraftFour/Type.php [hunk 2]
```diff
@@ -13,6 +13,26 @@
     const KEYWORD = 'type';
 
     /**
+     * Whether examples like 98249283749234923498293171823948729348710298301928331
+     * and "98249283749234923498293171823948729348710298301928331" are valid strings.
+     */
+    const BIGINT_MODE_STRING_VALID = 1;
```
Comment (yuloh) [inline src/Constraint/DraftFour/Type.php]: Since there are only 2 states, what about using a boolean?

```php
$type->allowLargeNumericStrings(true);
```

Not sure on the name.

Comment (yuloh) [inline src/Constraint/DraftFour/Type.php]: Maybe `largeNumericStringsAreValidStrings(true)`? Maybe a little verbose 😆

**Comment by yuloh** [comment]
Hey @coolhome 👋 

Long time no see 😁 Send me an email, IDK how to get a hold of you anymore.

**Comment by yuloh** [comment]
I'm tempted to just consider this a bug and fix it in a minor release.  As far as the validator knows it _is a string_, so it really shouldn't be rejecting it.

File: src/Constraint/DraftFour/Type.php (hunk 2 above)
Comment (akeeman) [inline src/Constraint/DraftFour/Type.php]: Boolean should do fine, indeed. And I think that is somewhat verbose too. Let us think of something for a while...

**Comment by akeeman** [comment]
As I see it, proposition option 1 can be seen as a bug fix, as it changes the behavior to be more desirable according to our logic. The current implementation doesn't change anything by default, but does add an extra option, so it's a new feature. If you intend to implement semver, then this should be released in a new feature release. :thinking:

**Comment by yuloh** [comment]
Oh right, I meant meant merging option 1 as a bug fix and releasing that.

I was originally thinking that it would be a breaking change to change the behavior, but the more I think about it no one should be relying on that behavior and I was wrong to add it in the first place.

Sorry for the back and forth!  I think you are right and getting rid of this will be better for everyone 👍

**Comment by yuloh** [comment]
I just tagged 1.0.1 which removes this.  Thanks for all of your help figuring this out and coming up with solutions.

**Comment by akeeman** [comment]
I actually am thinking for a while it would be better to use the configuration option to change the behavior then changing the default behavior, as it now diverges from the tests at https://github.com/json-schema/JSON-Schema-Test-Suite but you still advertise with "Passes the entire draft 4 JSON Schema Test Suite."

**Comment by yuloh** [comment]
Yeah, I will need to remove that claim from the docs I guess :(

If I preserved the behavior, it will break if they add `bignum string is still a string`, so I'm not sure if it's possible to pass 100% of the test suite and the optional tests in PHP.

**Comment by akeeman** [comment]
It is if the default behavior is in line with the test suite, but the other behavior is configurable.
Still it's worth a shot to contact the people from that test suite and check why this test is like that.

**Comment by akeeman** [comment]
And I was thinking, shouldn't the removal of such claim call for a new major version? :thinking:

**Comment by yuloh** [comment]
I felt a little unsure about releasing it as a bug fix, but here is why I think it qualifies as a bug:

Before the change, this would fail:

```
(new League\JsonGuard\Validator('9223372036854775999', json_decode('{"type": "string"}')))->passes(); // returns false, even though it's a string
```

...which means it's failing to validate according to the draft 4 specification.  It only passed the tests before because there isn't a test case for that specific scenario.

The 'bignum' test is optional -  it isn't actually a required part of the spec.  Since we can't pass that test without causing a bug that breaks compliance with the official spec, I think it's OK to fix the bug and skip that test.  That test was essentially relying on a bug to pass 😁 .

Changing it technically means that a schema containing the integer 9223372036854775999 decoded with JSON_BIGINT_AS_STRING will pass `{"type": "string"}` when it didn't before.  I think that is OK, because as far as the library is concerned it should pass validation; there is no definitive way for the constraint to know the string used to be an integer.

I would rather fix the legitimate bug (strings being rejected when they shouldn't) then keep a bug around because someone might be relying on it 😁 .

**Comment by akeeman** [comment]
I agree again, and apart from the test suite I couldn't agree more (so there you have the original idea behind this pr :wink: ). To support the "it's a bug" claim I've looked at the[ json schema specification on how stuff should be validated](http://json-schema.org/latest/json-schema-validation.html). There I can find stuff on why arbitrarily large integers should be supported, and thus the php implementation [should](http://json-schema.org/latest/json-schema-validation.html#rfc.section.3.2) support numeric strings as valid numbers. However, there's [no such thing](http://json-schema.org/latest/json-schema-validation.html#rfc.section.3.1) in there that I can find that disallows large numeric strings to be strings. So that's in line with our logic. In turn, that means that we can mark this legitimately as a bug.
As the draft and some logic supports our claim that it's a bug, it would be a good thing to fix it in the test suite too. If the maintainers agree and fix it, then this library again "Passes the entire draft 4 JSON Schema Test Suite.", which is a very important claim to me. To get this done, I'll file a pr there too (if not done already by someone else) and propose to remove the "a bignum is not a string" test.

So, to summarize: the claim that it's a bug is valid. It's just a bug in the test suite too. I'll file a pr there too.
